programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -28,43 +88,68 @@ struct CallerAccount<'a> {
     // mapped inside the vm (see serialize_parameters() in
     // BpfExecutor::execute).
     //
-    // When direct mapping is off, this includes both the account data _and_ the
-    // realloc padding. When direct mapping is on, account data is mapped in its
-    // own separate memory region which is directly mutated from inside the vm,
-    // and the serialized buffer includes only the realloc padding.
+    // This is only set when direct mapping is off (see the relevant comment in
+    // CallerAccount::from_account_info).
     serialized_data: &'a mut [u8],
     // Given the corresponding input AccountInfo::data, vm_data_addr points to
     // the pointer field and ref_to_len_in_vm points to the length field.
     vm_data_addr: u64,
-    ref_to_len_in_vm: &'a mut u64,
+    ref_to_len_in_vm: VmValue<'b, 'a, u64>,
     // To be removed once `feature_set::move_serialized_len_ptr_in_cpi` is active everywhere
     serialized_len_ptr: *mut u64,
     executable: bool,
     rent_epoch: u64,
 }
 
-impl<'a> CallerAccount<'a> {
+impl<'a, 'b> CallerAccount<'a, 'b> {
     // Create a CallerAccount given an AccountInfo.
     fn from_account_info(
         invoke_context: &InvokeContext,
-        memory_mapping: &MemoryMapping,
-        is_loader_deprecated: bool,
+        memory_mapping: &'b MemoryMapping<'a>,
         _vm_addr: u64,
         account_info: &AccountInfo,
-        original_data_len: usize,
-    ) -> Result<CallerAccount<'a>, Error> {
+        account_metadata: &SerializedAccountMetadata,
+    ) -> Result<CallerAccount<'a, 'b>, Error> {
+        let direct_mapping = invoke_context
+            .feature_set
+            .is_active(&feature_set::bpf_account_data_direct_mapping::id());
+
+        if direct_mapping {
+            check_account_info_pointer(
+                invoke_context,
+                account_info.key as *const _ as u64,
+                account_metadata.vm_key_addr,
+                "key",
+            )?;
+            check_account_info_pointer(
+                invoke_context,
+                account_info.owner as *const _ as u64,
+                account_metadata.vm_owner_addr,
+                "owner",
+            )?;
+        }
+
+        let original_data_len = account_metadata.original_data_len;
```

Comment:
nit: seems this variable is only used at the very last of just before returning from this fn. maybe, this var can be removed altogether?